scripts/gen-index.cts

```diff
@@ -0,0 +1,148 @@
+const PACKAGE_NAME = process.argv.at(WRITE_MODE ? 3 : 2);
+
+function isExported(node: ts.Declaration): boolean {
+	return (ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export) > 0;
+}
+
+function isType(node: ts.Node): boolean {
+	return [ts.SyntaxKind.InterfaceDeclaration, ts.SyntaxKind.TypeAliasDeclaration].includes(node.kind);
+}
+
+interface ModuleExportNodes {
+	normal: ts.Declaration[];
+	types: ts.Declaration[];
+}
+
+interface ModuleExportInclusions {
+	useModule: boolean;
+	useNormal: boolean;
+	useTypes: boolean;
+}
+
+class ModuleFile {
+	public readonly exports: ModuleExportNodes;
+	public readonly isPrivate: boolean;
+	public readonly exportInclusions: ModuleExportInclusions;
+	public readonly path: ParsedPath;
+	public readonly sourceFile: ts.SourceFile;
+
+	public constructor(sourceFile: ts.SourceFile) {
+		this.sourceFile = sourceFile;
+		this.path = parse(sourceFile.fileName);
+		this.isPrivate = this.path.name.startsWith('_');
+		this.exports = this.parseExports();
+		this.exportInclusions = this.deriveExportInclusions();
+	}
+
+	public generateExportSpecifiers(which: keyof ModuleExportNodes) {
+		return [...new Set(this.exports[which].map((node) => ts.getNameOfDeclaration(node)!.getText(this.sourceFile)!))].map((name) =>
+			ts.factory.createExportSpecifier(false, undefined, name)
+		);
+	}
+
+	/**
+	 * Rules:
+	 * 1. If the module begins with `_`, ignore it
+	 *    a) unless the module contains types, in which case those exclusively should be included
+	 * 2. If the module exclusively contains types, use `export type *`, otherwise use `export *`
+	 **/
+	public toExportDeclaration(packageDir: string): ts.ExportDeclaration | undefined {
+		const { useModule, useNormal, useTypes } = this.exportInclusions;
+		if (!useModule) return;
+
+		// TODO: make this more efficient
+		const typeExportSpecifiers = this.generateExportSpecifiers('types');
+		const exportSpecifiers = typeExportSpecifiers.concat(useNormal ? this.generateExportSpecifiers('normal') : []);
+		const relativePath = relative(packageDir, this.path.dir);
+		const adjustedPath = (this.path.base === 'index.ts' ? relativePath : join(relativePath, this.path.name)).split(winSep).join(posixSep);
+
+		return ts.factory.createExportDeclaration(
+			undefined,
+			!useNormal && useTypes,
+			this.isPrivate ? ts.factory.createNamedExports(exportSpecifiers) : undefined,
+			ts.factory.createStringLiteral(`./${adjustedPath}`, true),
+			undefined
+		);
+	}
+
+	private deriveExportInclusions(): ModuleExportInclusions {
+		const typesAvailable = this.exports.types.length > 0;
+
+		return {
+			useModule: !this.isPrivate || typesAvailable, // rule 1 & 1a
+			useNormal: !this.isPrivate && this.exports.normal.length > 0,
+			useTypes: typesAvailable
+		};
+	}
+
+	private parseExports(): ModuleExportNodes {
+		const normal: ts.Declaration[] = [];
+		const types: ts.Declaration[] = [];
+
+		// TODO: skip normals where this.isPrivate?
```

Comment:
Resolve in this PR, or delete the TODO. Your call. Adding a random TODO is poor management of technical debt. Taking up tech. debt is something we only do if there is a good reason to do so.

Reply:
upon further analysis doing that would require calling `isType` anyway so there would be little benefit, it would merely avoid unnecessary allocations, which doesn't really matter in a small script. removing shortly